src/Workspaces/Remote/ServiceHub/CodeLensOOP/ReferenceCodeLensProvider.cs

```diff
@@ -0,0 +1,228 @@
+
+            var dataPoint = new DataPoint(descriptor, await GetConnectionAsync(cancellationToken).ConfigureAwait(false), maxResult, projectIdGuid);
+            await dataPoint.TrackChangesAsync(cancellationToken).ConfigureAwait(false);
+
+            return dataPoint;
+        }
+
+        private async Task<Stream> GetConnectionAsync(CancellationToken cancellationToken)
+        {
+            // any exception from this will be caught by codelens engine and saved to log file and ignored.
+            // this follows existing code lens behavior and user experience on failure is owned by codelens engine
+            var callbackRpc = _codeLensCallbackService.GetCallbackJsonRpc(this);
+            var hostGroupId = await callbackRpc.InvokeWithCancellationAsync<string>("GetHostGroupIdAsync", arguments: null, cancellationToken).ConfigureAwait(false);
+
+            var hostGroup = new HostGroup(hostGroupId);
+            var serviceDescriptor = new ServiceDescriptor(RoslynCodeAnalysis) { HostGroup = hostGroup };
+
+            return await _client.RequestServiceAsync(serviceDescriptor, cancellationToken).ConfigureAwait(false);
+        }
+
+        private class DataPoint : IAsyncCodeLensDataPoint, IDisposable
+        {
+            private readonly JsonRpc _rpc;
+            private readonly int _maxResult;
+            private readonly Guid _projectIdGuid;
+
+            public DataPoint(CodeLensDescriptor descriptor, Stream stream, int maxResult, Guid projectIdGuid)
+            {
+                this.Descriptor = descriptor;
+
+                _rpc = new JsonRpc(new JsonRpcMessageHandler(stream, stream), target: this);
+                _rpc.JsonSerializer.Converters.Add(AggregateJsonConverter.Instance);
+
+                _maxResult = maxResult;
+                _projectIdGuid = projectIdGuid;
+
+                _rpc.StartListening();
+            }
+
+            public event AsyncEventHandler InvalidatedAsync;
+
+            public CodeLensDescriptor Descriptor { get; }
+
+            public async Task<CodeLensDataPointDescriptor> GetDataAsync(CancellationToken cancellationToken)
+            {
+                var referenceCount = await _rpc.InvokeWithCancellationAsync<ReferenceCount>(
+                    nameof(IRemoteCodeLensReferencesFromPrimaryWorkspaceService.GetReferenceCountAsync),
+                    new object[] { _projectIdGuid, Descriptor.FilePath, Descriptor.ApplicableToSpan.Value.ToTextSpan(), _maxResult }, cancellationToken).ConfigureAwait(false);
+
+                var referenceCountString = $"{referenceCount.Count}{(referenceCount.IsCapped ? "+" : string.Empty)}";
+
+                return new CodeLensDataPointDescriptor()
+                {
+                    Description = referenceCount.Count == 1 ?
+                                    string.Format(ServiceHubResources._0_reference, referenceCountString) :
+                                    string.Format(ServiceHubResources._0_references, referenceCountString),
```

Comment:
IDE style is to place ? and : on the next lines, indented once.

Reply:
sure